### PR TITLE
[vcpkg-tool] Allow overriding `FETCHCONTENT_FULLY_DISCONNECTED`

### LINKED ISSUE #28386: Forcing FETCHCONTENT_FULLY_DISCONNECTED is inappropriate and breaks projects using private registries
**Describe the bug**

#26959 added a change which forced setting the `FETCHCONTENT_FULLY_DISCONNECTED` CMake variable to true when building any port. This breaks ports in private vcpkg registries that rely on using FetchContent to pull in private things that are not available as vcpkg ports (I am working with a client doing exactly that in a production setting, and for good reasons).

While I understand that philosophically vcpkg would like all dependencies to be provided through vcpkg alone, there are plenty of scenarios where that is simply not appropriate and perhaps not even possible (for a variety of reasons). I'll give one example, but it is certainly not the only case.

An increasingly common scenario is where a company collects common build-related logic in a separate git repository and shares that among various projects. This may include toolchain files, so it has to be obtained before the first `project()` call. The common git repo may also set up company-specific things for vcpkg, such as environment variables that need to be defined to access internal servers, locations of private registries, and so on. The canonical pattern for this can be condensed down to something like this:

```cmake
cmake_minimum_required(VERSION 3.24)

include(FetchContent)
FetchContent_Declare(myCompCommonStuff
    GIT_REPOSITORY git@blah.example.com:/some/thing.git
    GIT_TAG        abc123abc123abc123...
)
FetchContent_MakeAvailable(myCompCommonStuff)

project(CoolProject)

do_standard_company_setup_things()

# Add stuff to the project as usual...
```

With the change in #26959, this now breaks when such a project is being made available as a port. The call to `FetchContent_MakeAvailable()` will turn into a no-op, so all of the things the project expects to have been defined by that separate repo (like the `do_standard_company_setup_things()` command in the above example) won't exist.

I disagree with the discussion towards the end of #26959. Expecting projects to duplicate their already working logic by doing things like calling `vcpkg_download_distfile()` or other methods (because things might not be available as a straight archive download) is unreasonable. If you want to force it for the public vcpkg repository, that's your call, but it is highly inappropriate to force it on private registries where the constraints and requirements can be very different. Company projects may be more than happy to not have some things stored in the asset cache (and in some security-sensitive scenarios, might _require_ that they not be cached). The choice should not be taken out of the hands of the maintainer of private registry ports.

As a CMake maintainer and author of FetchContent, I'll also add that I view this as an abuse of `FETCHCONTENT_FULLY_DISCONNECTED`. That option is intended for people who need to work offline, but who would still first need to be online to populate the dependencies initially. Using it to prevent users from populating any dependencies through FetchContent at all is a pretty aggressive move. I'd urge you to reconsider and revert that change.

**Environment**
- OS: all
- Compiler: all

**To Reproduce**

See example given above

**Expected behavior**

`FetchContent_MakeAvailable()` calls should continue to work, at the very least for people maintaining private registries.

**Failure logs**
N/A

**Additional context**

### PR DESCRIPTION
**Pull request by Thomas1664**
**Describe the pull request**

- #### What does your PR fix?
  Fixes #28386
  Alternative to #28483

We can't simply remove `FETCHCONTENT_FULLY_DISCONNECTED` because this would lead to bugs for users that want to build on isolated machines. Therefore, we have to set this variable in `vcpkg_configure_cmake()` as well. Since using `vcpkg_configure_cmake()` is deprecated, it shouldn't break ports in private registries. If a user still needs `FETCHCONTENT_FULLY_DISCONNECTED` set to `OFF`, they can override it via `OPTIONS`.

### CODE REVIEW

File: ports/vcpkg-cmake/vcpkg_cmake_configure.cmake [hunk 1]
```diff
@@ -195,6 +194,10 @@ function(vcpkg_cmake_configure)
         endif()
     endforeach()
 
+    if(NOT "-DFETCHCONTENT_FULLY_DISCONNECTED=OFF" IN_LIST arg_OPTIONS)
```
Comment (dg0yt) [inline ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:197]: No, this is too tight. Users might use `:BOOLEAN` or `=0` etc.

File: ports/vcpkg-cmake/vcpkg_cmake_configure.cmake [hunk 2]
```diff
@@ -195,6 +194,10 @@ function(vcpkg_cmake_configure)
...
+        vcpkg_list(APPEND arg_OPTIONS "-DFETCHCONTENT_FULLY_DISCONNECTED=ON")
+    endif()
```
Comment (dg0yt) [inline ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:199]: All we need is to *prepend* this option, without condition. Then a user-supplied option will override our default.

**Review by dg0yt — CHANGES_REQUESTED** [review]

**Comment by ras0219-msft** [comment]
While `vcpkg_configure_cmake()` is deprecated, that does not mean we can break existing ports. We must take #28483, which reverts the change only for `vcpkg_configure_cmake()`.

However, we absolutely should support the override for `vcpkg_cmake_configure()` as you've done in this PR. I think @dg0yt's suggested improvement is best -- we need to move the vcpkg "preset" options above the user's supplied `OPTIONS`.

**Comment by Neumann-A** [comment]
`VCPKG_CMAKE_CONFIGURE_OPTIONS` is designed to override all settings. So I don't get why this is even needed.

**Comment by dg0yt** [comment]
> `VCPKG_CMAKE_CONFIGURE_OPTIONS` is designed to override all settings. So I don't get why this is even needed.

IMO it is generally bad to override anything which is already passed explicitly in `OPTIONS`. Because this is the primary interface for this purpose.
`VCPKG_CMAKE_CONFIGURE_OPTIONS` is nice to have for use in triplet files.

**Comment by Thomas1664** [comment]
> We must take #28483, which reverts the change only for `vcpkg_configure_cmake()`.

Why do we have to take #28483 before merging this PR? This would introduce a bug (see PR description above) until this PR gets merged.

**Comment by Cheney-W** [comment]
@Thomas1664 Could you please handle the conflicts?

**Review by Cheney-W — DISMISSED** [review]

**Comment by BillyONeal** [comment]
Thanks for your fix :). I merged Robert's change which had only baseline issues, merged that with this, and committed (as this has a green run).

**Comment by Thomas1664** [comment]
> Thanks for your fix :). I merged Robert's change which had only baseline issues, merged that with this, and committed (as this has a green run).

@BillyONeal Now `FETCHCONTENT_FULLY_DISCONNECTED` is not enabled in `vcpkg_configure_cmake`. What's wrong with setting it on this one as well? one can override it just like with `vcpkg_cmake_configure`.

**Comment by BillyONeal** [comment]
> What's wrong with setting it on this one as well?

It breaks backwards compatibility with old port versions.

**Pull request merged**